panther_hardware_interfaces/src/panther_system_ros_interface.cpp

```
@@ -0,0 +1,128 @@
+
+  driver_state.rear.fault_flag = rear.GetFaultFlag().GetMessage();
+  driver_state.rear.script_flag = rear.GetScriptFlag().GetMessage();
+  driver_state.rear.left_motor_runtime_error = rear.GetLeftRuntimeError().GetMessage();
+  driver_state.rear.right_motor_runtime_error = rear.GetRightRuntimeError().GetMessage();
+}
+
+void PantherSystemRosInterface::UpdateMsgDriversParameters(
+  const DriverState & front, const DriverState & rear)
+{
+  auto & driver_state = realtime_driver_state_publisher_->msg_;
+
+  driver_state.front.voltage = front.GetVoltage();
+  driver_state.front.current = front.GetCurrent();
+  driver_state.front.temperature = front.GetTemperature();
+
+  driver_state.rear.voltage = rear.GetVoltage();
+  driver_state.rear.current = rear.GetCurrent();
+  driver_state.rear.temperature = rear.GetTemperature();
+}
+
+void PantherSystemRosInterface::UpdateMsgErrors(const CanErrors & can_errors)
+{
+  auto & driver_state = realtime_driver_state_publisher_->msg_;
+
+  driver_state.error = can_errors.error;
+  driver_state.write_sdo_error = can_errors.write_sdo_error;
+  driver_state.read_sdo_error = can_errors.read_sdo_error;
+  driver_state.read_pdo_error = can_errors.read_pdo_error;
+
+  driver_state.front.data_timed_out = can_errors.front_data_timed_out;
+  driver_state.rear.data_timed_out = can_errors.rear_data_timed_out;
+
+  driver_state.front.can_net_err = can_errors.front_can_net_err;
+  driver_state.rear.can_net_err = can_errors.rear_can_net_err;
+}
+
+void PantherSystemRosInterface::PublishDriverState()
+{
+  if (realtime_driver_state_publisher_->trylock()) {
+    realtime_driver_state_publisher_->unlockAndPublish();
+  }
+}
+
+void PantherSystemRosInterface::ClearErrorsCb(
+  std_srvs::srv::Trigger::Request::ConstSharedPtr /* request */,
+  std_srvs::srv::Trigger::Response::SharedPtr response)
+{
+  RCLCPP_INFO(rclcpp::get_logger("PantherSystem"), "Clearing errors");
+  clear_errors_();
+  response->success = true;
```

Service could be put inside try, catch block

---

I'm not sure which exception should it catch. This function won't throw any, so can't be anything specific. I think that we talked about it, and specified that it should catch all exceptions, but in this case it should be also changed in the `SetBrightnessCB` (the only other service that we currently have).

---

The idea was that service that does not work correctly wont kill the node in case it is badly implemented. I agree that, for simple functions there is no point in doing this.

---

In this case, I would prefer to add try catch always (so that we won't have to analyze if we call simple function or not). We have to remember to also adjust it in the `SetBrightnessCB`.